build.zig

```diff
@@ -874,6 +874,8 @@ fn maybe_execute(
     binary_name: []const u8,
 ) void {
     var to_run = std.ArrayList([]const u8).init(allocator);
+    defer to_run.deinit();
```

Comment:
I would rather prefer us to have a habit of always doing defer/deinit. This way the code is correct if it's ever moved to a new context or switched to another allocator.

Reply:
> I would say that such inconsistency is a natural consequence of trying to always init/deinit

Agreed. On a related note (but not related to the Build.zig example), I do like your thinking on create/destroy etc. and we should use that where appropriate. I don't actually mean that we must always only use `init/deinit()`.

> So I'd say a more robust strategy is to avoid writing generic allocating code --- either allocate all memory at startup, or figure out how to use arenas to do bulk deallocation.

Agreed. However, there are parts of our code, which may up in different contexts outside of TigerBeetle, which we can't anticipate now. For example, clients that we expose in different languages, where we may not control the allocator policy. Or embedded TigerBeetle mode... in a not-to-distant galaxy.

So what I meant is that wherever we allocate internally, we _must_ make a habit of defer'ing or errdefer'ing the correct deallocation code, without breaking encapsulation and assuming anything about the allocator. I have always thought of this as good hygiene, to build muscle around thinking about deallocations, and when reviewing, but the "don't lose flexibility" is the big one.

And again, being strict about deallocation also doesn't preclude these defense-in-depth strategies such as static allocation or arenas. We can do both, but I would like us to write the code assuming it needs to be correct regardless of allocator or context.

As far as possible, we should allow the user to set the allocator policy externally to the implementation, and prefer unmanaged interfaces. This preserves flexibility and minimizes contract.

At the same time, where it makes sense, we can be coupled to allocator strategy, for example to use the arena trick for resetting the storage engine.

> Like, it's not hard to get 9 out of 10 defers right, but getting them all is something we are not particularly good at.

“Do not go gentle into that good night!”

For example, this is why TigerStyle encourages a newline after allocation, so that the reviewer can more easily check that the corresponding deallocation is there.

I think we could also leverage Zig's test allocator more to get comfort here?

Given that our outer policy is also static allocation, we should also be able to detect leaks ourselves by measuring any memory increase through repeated resets?